### ISSUE TITLE
/aoi_s/aoi_supp/trampoline: line 5: (...): No such file or directory

### ISSUE THREAD

**Issue by probonopd** [lvml/makeaoi#7]
Made an AppImage on my local machine which I then wanted to run on [Travis CI](https://travis-ci.org/).

Getting:

```
/aoi_supp/trampoline: line 5: /home/travis/build/AppImage/AppImageHub/appstreamcli: No such file or directory
```

Reference:
https://travis-ci.org/AppImage/AppImageHub/builds/266456025#L995

This is how I made the AppImage locally:

```
sudo apt-get install -yq --no-install-recommends tcl cmake gettext itstool libglib2.0-dev libxml2-dev gtk-doc-tools libgirepository1.0-dev qt5-default libyaml-dev xmlto publican gobject-introspection libstemmer-dev valac git
git clone https://github.com/ximion/appstream
cd appstream/
mkdir build
cd build/
cmake .. -DQT=OFF -DVAPI=OFF -DDOCUMENTATION=OFF -DMAINTAINER=OFF -DSTEMMING=OFF -DAPT_SUPPORT=OFF
make
git clone https://github.com/lvml/makeaoi
cd makeaoi
make
cd ..
./makeaoi/makeaoi trace foo ./tools/appstreamcli /usr/share/appdata/catfish.appdata.xml 
./makeaoi/makeaoi populate foo
rm -rf foo/root_overlay/usr/share/appdata/
wget https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage
chmod a+x appimagetool-x86_64.AppImage 
./appimagetool-x86_64.AppImage --no-appstream foo
```

This is how I "install" it on Travis CI:

```
sudo wget -c "https://github.com/AppImage/AppImageHub/releases/download/deps/appstreamcli-x86_64.AppImage" -O /usr/local/bin/appstreamcli
sudo chmod a+x /usr/local/bin/appstreamcli
```

**Comment by probonopd** [comment]
I think it doesn't bundle the main executable. Because it is not in `/usr/bin/`?

**Comment by probonopd** [comment]
Looks like the main executable does get packaged when I do it like this:

```
sudo apt-get install -yq --no-install-recommends tcl cmake gettext itstool libglib2.0-dev libxml2-dev gtk-doc-tools libgirepository1.0-dev qt5-default libyaml-dev xmlto publican gobject-introspection libstemmer-dev valac git
git clone https://github.com/ximion/appstream
cd appstream/
mkdir build
cd build/
cmake .. -DQT=OFF -DVAPI=OFF -DDOCUMENTATION=OFF -DMAINTAINER=OFF -DSTEMMING=OFF -DAPT_SUPPORT=OFF -DCMAKE_INSTALL_PREFIX=/usr
make
sudo make install
git clone https://github.com/lvml/makeaoi
cd makeaoi
make
cd ..
./makeaoi/makeaoi trace foo /usr/bin/appstreamcli /usr/share/appdata/catfish.appdata.xml 
./makeaoi/makeaoi populate foo
rm -rf foo/root_overlay/usr/share/appdata/
wget https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage
chmod a+x appimagetool-x86_64.AppImage 
./appimagetool-x86_64.AppImage --no-appstream foo
```

Is this intentional?

**Comment by probonopd** [comment]
But even then the problem persists. Even on my local Ubuntu machine, it stops working as soon as I put the AppImage into a different directory, cd there, and try to run it from there.

```
me@host:~/appstream/build$ cp appstreamcli-x86_64.AppImage /home/me/
me@host:~$ cd /home/me/
me@host:~$ ./appstreamcli-x86_64.AppImage 
/aoi_supp/trampoline: line 5: /home/me/tools/appstreamcli: No such file or directory
```

**Comment by lvml** [comment]
The /home/ directory is explicitely excluded by default in makeaoi - see https://github.com/lvml/makeaoi/blob/master/makeaoi.tcl#L168

This is because usually, applications are working with user-specific data- or configuration-files that you don't want to include in an overlay image - and also usually, the (non-user-specific) applications themselves are not residing below /home.

However, if you like to use makeaoi on an application residing in /home on the source machine, just remove the exclusion at https://github.com/lvml/makeaoi/blob/master/makeaoi.tcl#L170
If you do, just make sure you exclude unwanted data or configuration files from /home/somwhere manually when editing packlist.txt

**Comment by probonopd** [comment]
Wouldn't it make sense to _always_ include the main executable (`$0`) regardless of where it is stored? I mean, it's safe to assume that it will be needed...

**Comment by probonopd** [comment]
But then, as I wrote above, the problem persists when I move the main executable to `/usr/bin` (and have made sure it ends up in the bundle).

**Comment by probonopd** [comment]
As a semi-OT side note, check this AppImage that bundles _everything_ including the loader, but does so without any external (static-*) components, just by setting rpaths and invoking the loader directly: https://github.com/AppImage/appimage.github.io/releases/download/deps/appstreamcli-x86_64.AppImage

**Comment by lvml** [comment]
> Wouldn't it make sense to always include the main executable ($0) regardless of where it is stored?

While such would certainly be trivial to implement, chances are near 100% that this executable would also require other files out of the excluded /home/ directory - if it was a "just one exectuable file" application, people would probably not make use of makeaoi to ship it somewhere.

I implemented a change to abort with an explanatory error message if an executable is traced that resides in an excluded path.

**Comment by lvml** [comment]
Regarding:
> /aoi_supp/trampoline: line 5: /home/me/tools/appstreamcli: No such file or directory

I cannot reproduce this behaviour.
Please check where the symlink "exe_AppRun" (in the directory you prepared with makeaoi) points to. This is the location that the "trampoline" script expects to find the executable at.

**Comment by lvml** [comment]
Regarding:
> AppImage that bundles everything including the loader, but does so without any external (static-*) components, just by setting rpaths and invoking the loader directly

Setting rpaths only helps with finding shared libraries that have been linked to an executable, it does not help for objects opened with dlopen(), and of course it does not help for data files opened with open(). But many applications use dlopen() or require data files somewhere - starting with every application that uses glibc.

(BTW: Is this about https://www.freedesktop.org/software/appstream/docs/api/re22.html or a different piece of software?)

**Comment by probonopd** [comment]
Yes, it is about that software. The problem here is that this software changes according to ever-changing specifications and requires new versions of e.g., glib, yet I want to use it on (old) build systems...

**Comment by lvml** [comment]
Hmm... in addition to my concerns regarding reliance on rpath I must say that I see not too many similarities between appstream and makeaoi.
Also, I usually stop considering a tool when I read in its documentation that it expects humans to write XML-formatted configuration files.

**Comment by probonopd** [comment]
>  I see not too many similarities between appstream and makeaoi.

Nor do I. I just tried to "package" the `appstreamcli` tool using makeaoi. And I agree about the XML. But they seem to (partly) be [moving to YAML](https://wiki.debian.org/DEP-11).

**Comment by lvml** [comment]
Oh, seems I misunderstood your reference to appstreamcli.

Can you reproduce the "/aoi_supp/trampoline: line 5: /home/me/tools/appstreamcli: No such file or directory" issue possibly with something simple like /usr/bin/ls as the executable? Any idea where the "/home/me/" path may come from?

**Comment by probonopd** [comment]
`/home/me` was my `$CWD`.
Will try to reproduce this with another application.
